nexus/mgs-updates/src/rot_updater.rs

```diff
@@ -203,23 +210,218 @@ impl SpComponentUpdateHelper for ReconfiguratorRotUpdater {
     /// Checks if the component is already updated or ready for update
     fn precheck<'a>(
         &'a self,
-        _log: &'a slog::Logger,
-        _mgs_clients: &'a mut MgsClients,
-        _update: &'a PendingMgsUpdate,
+        log: &'a slog::Logger,
+        mgs_clients: &'a mut MgsClients,
+        update: &'a PendingMgsUpdate,
     ) -> BoxFuture<'a, Result<PrecheckStatus, PrecheckError>> {
-        // TODO-K: fill in the precheck
-        todo!()
+        async move {
+            // Verify that the device is the one we think it is.
+            let state = mgs_clients
+            .try_all_serially(log, move |mgs_client| async move {
+                mgs_client.sp_get(update.sp_type, update.slot_id).await
+            })
+            .await?
+            .into_inner();
+            debug!(log, "found SP state"; "state" => ?state);
+            if state.model != update.baseboard_id.part_number
+                || state.serial_number != update.baseboard_id.serial_number
+            {
+                return Err(PrecheckError::WrongDevice {
+                    sp_type: update.sp_type,
+                    slot_id: update.slot_id,
+                    expected_part: update.baseboard_id.part_number.clone(),
+                    expected_serial: update.baseboard_id.serial_number.clone(),
+                    found_part: state.model,
+                    found_serial: state.serial_number,
+                });
+            }
+
+            let PendingMgsUpdateDetails::Rot {
+                expected_slot_a_version,
+                expected_slot_b_version,
+                expected_active_slot,
+                expected_persistent_boot_preference,
+                ..
+            } = &update.details
+            else {
+                unreachable!(
+                    "pending MGS update details within ReconfiguratorRotUpdater \
+                    will always be for the RoT"
+                );
+            };
+
+            // Fetch the caboose from the currently active slot.
+            let caboose = mgs_clients
+            .try_all_serially(log, move |mgs_client| async move {
+                mgs_client
+                    .sp_component_caboose_get(
+                        update.sp_type,
+                        update.slot_id,
+                        &SpComponent::ROT.to_string(),
+                        expected_active_slot.to_u16(),
+                    )
+                    .await
+            })
+            .await?
+            .into_inner();
+            debug!(log, "found active slot caboose"; "caboose" => ?caboose);
+
+            // If the version in the currently active slot matches the one we're
+            // trying to set, then there's nothing to do.
+            if caboose.version == update.artifact_version.as_str() {
+                return Ok(PrecheckStatus::UpdateComplete);
+            }
+
+            // Otherwise, if the version in the currently active slot does not
```

Comment:
These two expected version checks are fairly similar to what can be found in `sp_updater.rs` but adapted to fit the case where either slot could be the active one